taverna-cwl-activity-ui/src/main/java/org/apache/taverna/cwl/ui/serviceprovider/CwlServiceProvider.java

```diff
@@ -16,72 +16,167 @@
-	private File cwlFilesLocation;
+
+	private static final String providerName = "CWL Services";
+	private static final URI providerId = CwlServiceDesc.ACTIVITY_TYPE.resolve("#provider");
 
 	@Override
 	public void findServiceDescriptionsAsync(FindServiceDescriptionsCallBack callBack) {
 
 		// get the location of the cwl tool from the workbench
-		cwlFilesLocation = new File(getConfiguration().getPath());
-		// This is holding the CWL configuration beans
-		List<CwlServiceDesc> result = new ArrayList<CwlServiceDesc>();
-
-		File[] cwlFiles = getCwlFiles();
-
-		// Load the CWL file using SnakeYaml lib
-		Yaml cwlReader = new Yaml();
+		ArrayList<Path> paths = getPath();
 
-		for (File file : cwlFiles) {
-			Map cwlFile = null;
+		for (Path path : paths) {
+			// figure out the dots in the path ex: /maanadev/../cwltools
+			Path normalizedPath = path.normalize();
 
+			DirectoryStream<Path> stream = null;
 			try {
-				cwlFile = (Map) cwlReader.load(new FileInputStream(file));
-			} catch (FileNotFoundException e) {
-				e.printStackTrace();
-			}
-			if (cwlFile != null) {
-				// Creating CWl service Description
-				CwlServiceDesc cwlServiceDesc = new CwlServiceDesc();
-				cwlServiceDesc.setCwlConfiguration(cwlFile);
-				cwlServiceDesc.setToolName(file.getName().split("\\.")[0]);
-
-				// add to the result
-				result.add(cwlServiceDesc);
-				// return the service description
-				callBack.partialResults(result);
+				stream = Files.newDirectoryStream(normalizedPath, "*.cwl");
+			} catch (IOException e) {
+				logger.warn("Path is not correct !");
+				return;
 			}
+			// create stream with parallel capabilities
+			Stream<Path> paralleStream = StreamSupport.stream(stream.spliterator(), true);
+
+			paralleStream.forEach(p -> {
+				Yaml reader = getYamlReader();
+
+				Map cwlFile;
+				try (FileInputStream file = new FileInputStream(path.toFile())) {
+					cwlFile = (Map) reader.load(file);
+					JsonNode config = createJsonNode(p, cwlFile);
+					// Creating CWl service Description
+					CwlServiceDesc cwlServiceDesc = createCWLDesc(config);
+					// return the service description
+					callBack.partialResults(Arrays.asList(cwlServiceDesc));
+
+				} catch (IOException e) {
+
+					logger.warn("File not Found !");
 
+				}
+
+			});
+
+			callBack.finished();
 		}
-		callBack.finished();
 
 	}
+/**
+ * This method checks whether provided path is valid or not and if it's valid the it's added to the list
+ * @param defaultPaths arrylist to hold valid paths
+ * @param path 
+ * @param path1 if there is no second path argument this should be set to null
+ */
+	public void addPath(ArrayList<Path> defaultPaths, String path, String path1) {
+
+		Path defaultPath;
+		if (path1 == null)
+			defaultPath = Paths.get(path);
+		else
+			defaultPath = Paths.get(path, path1);
+
+		if (defaultPath.isAbsolute())
+			defaultPaths.add(defaultPath);
+	}
 
-	
+	private ArrayList<Path> getPath() {
```

Comment:
Return as interface `List<Path>` instead of hard-coding `ArrayList` - we don't need ArrayList-specific features.